pilot/pkg/networking/core/v1alpha3/cluster.go

```diff
@@ -178,10 +178,55 @@ func (configgen *ConfigGeneratorImpl) buildOutboundClusters(proxy *model.Proxy,
 		Node: proxy,
 	}
 	networkView := model.GetNetworkView(proxy)
+	tmpPush := *push
+
+	if proxy.Type == model.Router {
+		tmpNode := *proxy
+		log.Infof("Collecting clusters names and route names for the node: %v", proxy.ID)
+		tmpPush.GatewayClusterNames = map[string]bool{}
+		rawListeners := configgen.BuildListeners(&tmpNode, &tmpPush)
+		for _, l := range rawListeners {
+			if err := l.Validate(); err != nil {
+				log.Warnf("CDS: Generated invalid listener for node:%s: %v, %v. Falling back to sending unfiltered cluster config", proxy.ID, err, l)
+				// Generating invalid listeners is a bug.
+				// Instead of panic, which will break down the whole cluster. Just ignore it here an  d let CDS push all the clusters without filtering for this gateway proxy
+				tmpPush.GatewayClusterNames = map[string]bool{}
+				break
+			}
+		}
+		if len(tmpPush.GatewayRouteNames) != 0 {
+			log.Debugf("Cached %v routenames : %v", len(tmpPush.GatewayRouteNames), tmpPush.GatewayRouteNames)
+			rawRoutes := configgen.BuildHTTPRoutes(&tmpNode, &tmpPush, tmpPush.GatewayRouteNames)
+			for _, r := range rawRoutes {
+				if err := r.Validate(); err != nil {
+					log.Warnf("CDS: Generated invalid route for node:%s: %v, %v. Falling back to sending unfiltered cluster config", proxy.ID, err, r)
+					// Generating invalid routes is a bug.
+					// Instead of panic, which will break down the whole cluster. Just ignore it here an  d let CDS push all the clusters without filtering for this gateway proxy
+					tmpPush.GatewayClusterNames = map[string]bool{}
+					break
+				}
+				for _, vh := range r.VirtualHosts {
+					for _, vhr := range vh.Routes {
+						ra := vhr.GetRoute()
+						if ra != nil && ra.GetCluster() != "" {
```

Comment:
per offline feedback, store the cluster names in model.Proxy struct. Second, instead of parsing the routes from each virtual host, try to store the clusters as you create the routes in route/route.go . We have weightedClusters, non weightedClusters, mirrored clusters, and bunch of other things. So, i thikn your best bet is to track these in route/route.go

Reply:
The tricky point is: cluster is the first resource generated.

Reply:
@rshriram route/route.go can be used to track clusters generated by virtual services only. I think that Routes can also be patched/added using EnvoyFilter api and to track clusters in those routes, I think iterating over returned final list of virtual hosts is less error prone and less ugly as compared to sprinkling this cluster name collection logic all over the place. WDYT?

Reply:
Fine, that is a fair point but remember that EnvoyFilter can also have custom filters that refer to clusters and we have no way of knowing these clusters. Either way, this logic can be done after constructing all the routes for a gateway, and then constructing the clusters.